File: android/hummer-core/src/main/assets/HummerDefinition.js
const __GLOBAL__ = this;

const HUMMER_OBJECT_PREFIX = '-_-_-_hummer-object_-_-_-';
const HUMMER_ARRAY_PREFIX = '-_-_-_hummer-array_-_-_-';

var hm_id = 1;
const idGenerator = () => hm_id++;

const transSingleArg = (arg) => {
    if (arg instanceof Base) {
        arg = arg.objID;
    } else if (arg instanceof Function) {
        // do nothing
    } else if (arg instanceof Object) {
        arg = JSON.stringify(arg);
    }
    return arg;
}

const transSingleArgWithPrefix = (arg) => {
    if (arg instanceof Base) {
        arg = arg.objID;
    } else if (arg instanceof Function) {
        // do nothing
    } else if (arg instanceof Array) {
        arg = HUMMER_ARRAY_PREFIX + JSON.stringify(arg);
    } else if (arg instanceof Object) {
        arg = HUMMER_OBJECT_PREFIX + JSON.stringify(arg);
    }
    return arg;
}

const transArgs = (...args) => {
    for (let i = 0; i < args.length; i++) {
        args[i] = transSingleArg(args[i]);
    }
    return args;
}

const transArgsWithPrefix = (...args) => {
    for (let i = 0; i < args.length; i++) {
        args[i] = transSingleArgWithPrefix(args[i]);
    }
    return args;
}

const console = {
    log: (...msgs) => printLog("console.log", ...msgs),
    debug: (...msgs) => printLog("console.debug", ...msgs),
    info: (...msgs) => printLog("console.info", ...msgs),
    warn: (...msgs) => printLog("console.warn", ...msgs),
    error: (...msgs) => printLog("console.error", ...msgs)
}

var __IS_DEBUG__ = false;
/**
 * 下面的日志打印逻辑，是为了支持前端console.log的Object类型参数和多参数情况
 */
const printLog = (funcName, ...msgs) => {
    if (__IS_DEBUG__) {
        let msg = '';
        if (msgs.length == 1) {
            let m = msgs[0];
            if (m instanceof Function) {
                msg = m.toString();
            } else if (m instanceof Object) {
                msg = JSON.stringify(m);
            } else if (typeof m === 'undefined') {
                msg = 'undefined';
            } else if (typeof m === 'null') {
                msg = 'null';
            } else {
                msg = m.toString();
            }
        } else if (msgs.length > 1) {
            for (let i = 0; i < msgs.length; i++) {
                if (i > 0) {
                    msg = msg.concat(', ');
                }
                let m = msgs[i];
                if (m instanceof Function) {
                    msg = msg.concat(m.toString());
                } else if (m instanceof Object) {
                    msg = msg.concat(JSON.stringify(m));
                } else if (typeof m === 'undefined') {
                    msg = msg.concat('undefined');
                } else if (typeof m === 'null') {
                    msg = msg.concat('null');
                } else {
                    msg = msg.concat(m);
                }
            }
        }
        invoke("Hummer", 0, funcName, msg);
    }
}

const setTimeout = (func, timeout) => {
    let timer = new Timer();
    timer.setTimeout(func, timeout);
    return timer;
}

const clearTimeout = (timer) => {
    if (timer instanceof Timer) {
        timer.clearTimeout();
    }
}

const setInterval = (func, interval) => {
    let timer = new Timer();
    timer.setInterval(func, interval);
    return timer;
}

const clearInterval = (timer) => {
    if (timer instanceof Timer) {
        timer.clearInterval();
    }
}

const NotifyCenter = {
    addEventListener: (event, callback) => {
        invoke("NotifyCenter", 0, "addEventListener", event, callback);
    },
    removeEventListener: (event, callback) => {
        invoke("NotifyCenter", 0, "removeEventListener", event, callback);
    },
    triggerEvent: (event, value) => {
        invoke("NotifyCenter", 0, "triggerEvent", event, JSON.stringify(value));
    }
}

const Hummer = {
    setBasicWidth: (width) => {
        invoke("Hummer", 0, "setBasicWidth", width);
    },
    render: (view) => {
        invoke("Hummer", 0, "render", view.objID);
    },
    loadScript: (script) => {
        return invoke("Hummer", 0, "loadScript", script);
    },
    loadScriptWithUrl: (url, callback) => {
        invoke("Hummer", 0, "loadScriptWithUrl", url, callback);
    },
    postException: (err) => {
        err = transSingleArg(err);
        invoke("Hummer", 0, "postException", err);
    },
    notifyCenter: NotifyCenter,
}

class Base {
    constructor(className, ...args) {
        this.className = className;
        this.objID = idGenerator();
        this.recycler = new Recycler(this.objID);

        let params = transArgs(...args);
        invoke(this.className, this.objID, "constructor", this, ...params);

        // 已弃用
        this.initialize(...args);

        // 此方法只用于调试，为了统计组件树和函数调用树
        if (__IS_DEBUG__) {
            invoke(this.className, this.objID, "constructor_end", this);
        }
    }

    // 已弃用
    initialize(...args) {}

    set style(arg) {
        this._style = arg;
        arg = transSingleArg(arg);
        invoke(this.className, this.objID, "setStyle", arg);
    }

    get style() {
        return this._style;
    }

    set enabled(arg) {
        this._enabled = arg;
        invoke(this.className, this.objID, "setEnabled", arg);
    }

    get enabled() {
        return this._enabled;
    }

    set accessible(arg) {
        this._accessible = arg;
        invoke(this.className, this.objID, "setAccessible", arg);
    }

    get accessible() {
        return this._accessible;
    }

    set accessibilityLabel(arg) {
        this._accessibilityLabel = arg;
        invoke(this.className, this.objID, "setAccessibilityLabel", arg);
    }

    get accessibilityLabel() {
        return this._accessibilityLabel;
    }

    set accessibilityHint(arg) {
        this._accessibilityHint = arg;
        invoke(this.className, this.objID, "setAccessibilityHint", arg);
    }

    get accessibilityHint() {
        return this._accessibilityHint;
    }

    set accessibilityRole(arg) {
        this._accessibilityRole = arg;
        invoke(this.className, this.objID, "setAccessibilityRole", arg);
    }

    get accessibilityRole() {
        return this._accessibilityRole;
    }

    set accessibilityState(arg) {
        this._accessibilityState = arg;
        arg = transSingleArg(arg);
        invoke(this.className, this.objID, "setAccessibilityState", arg);
    }

    get accessibilityState() {
        return this._accessibilityState;
    }

    addEventListener(...args) {
        invoke(this.className, this.objID, "addEventListener", ...args);
    }

    removeEventListener(...args) {
        invoke(this.className, this.objID, "removeEventListener", ...args);
    }

    addAnimation(...args) {
        let stash = args;
        args = transArgs(...args);
        invoke(this.className, this.objID, "addAnimation", ...args);
    }

    removeAnimationForKey(arg) {
        invoke(this.className, this.objID, "removeAnimationForKey", arg);
    }

    removeAllAnimation() {
        invoke(this.className, this.objID, "removeAllAnimation");
    }

    getRect(arg) {
        invoke(this.className, this.objID, "getRect", arg);
    }

    resetStyle() {
        invoke(this.className, this.objID, "resetStyle");
    }

    recycle() {
        invoke(this.className, this.objID, "recycle");
    }
}

__GLOBAL__.Hummer = Hummer;
__GLOBAL__.Base = Base;

File: android/hummer-sdk/src/main/java/com/didi/hummer/context/HummerDefinition.java
package com.didi.hummer.context;

/**
 * Created by XiaoFeng on 2022/6/2.
 */
public class HummerDefinition {

    /**
     * 正常模式下的核心定义
     */
    public static final String CORE = "const __GLOBAL__ = this;\n" +
            "\n" +
            "const HUMMER_OBJECT_PREFIX = '-_-_-_hummer-object_-_-_-';\n" +
            "const HUMMER_ARRAY_PREFIX = '-_-_-_hummer-array_-_-_-';\n" +
            "\n" +
            "var hm_id = 1;\n" +
            "const idGenerator = () => hm_id++;\n" +
            "\n" +
            "const transSingleArg = (arg) => {\n" +
            "    if (arg instanceof Base) {\n" +
            "        arg = arg.objID;\n" +
            "    } else if (arg instanceof Function) {\n" +
            "        // do nothing\n" +
            "    } else if (arg instanceof Object) {\n" +
            "        arg = JSON.stringify(arg);\n" +
            "    }\n" +
            "    return arg;\n" +
            "}\n" +
            "\n" +
            "const transSingleArgWithPrefix = (arg) => {\n" +
            "    if (arg instanceof Base) {\n" +
            "        arg = arg.objID;\n" +
            "    } else if (arg instanceof Function) {\n" +
            "        // do nothing\n" +
            "    } else if (arg instanceof Array) {\n" +
            "        arg = HUMMER_ARRAY_PREFIX + JSON.stringify(arg);\n" +
            "    } else if (arg instanceof Object) {\n" +
            "        arg = HUMMER_OBJECT_PREFIX + JSON.stringify(arg);\n" +
            "    }\n" +
            "    return arg;\n" +
            "}\n" +
            "\n" +
            "const transArgs = (...args) => {\n" +
            "    for (let i = 0; i < args.length; i++) {\n" +
            "        args[i] = transSingleArg(args[i]);\n" +
            "    }\n" +
            "    return args;\n" +
            "}\n" +
            "\n" +
            "const transArgsWithPrefix = (...args) => {\n" +
            "    for (let i = 0; i < args.length; i++) {\n" +
            "        args[i] = transSingleArgWithPrefix(args[i]);\n" +
            "    }\n" +
            "    return args;\n" +
            "}\n" +
            "\n" +
            "const console = {\n" +
            "    log: (...msgs) => printLog(\"console.log\", ...msgs),\n" +
            "    debug: (...msgs) => printLog(\"console.debug\", ...msgs),\n" +
            "    info: (...msgs) => printLog(\"console.info\", ...msgs),\n" +
            "    warn: (...msgs) => printLog(\"console.warn\", ...msgs),\n" +
            "    error: (...msgs) => printLog(\"console.error\", ...msgs)\n" +
            "}\n" +
            "\n" +
            "var __IS_DEBUG__ = false;\n" +
            "/**\n" +
            " * 下面的日志打印逻辑，是为了支持前端console.log的Object类型参数和多参数情况\n" +
            " */\n" +
            "const printLog = (funcName, ...msgs) => {\n" +
            "    if (__IS_DEBUG__) {\n" +
            "        let msg = '';\n" +
            "        if (msgs.length == 1) {\n" +
            "            let m = msgs[0];\n" +
            "            if (typeof m === 'undefined') {\n" +
            "                msg = 'undefined';\n" +
            "            } else if (m == null) {\n" +
            "                msg = 'null';\n" +
            "            } else if (m instanceof Error) {\n" +
            "                msg = m.toString() + '\\n' + m.stack;\n" +
            "            } else if (m instanceof Function) {\n" +
            "                msg = m.toString();\n" +
            "            } else if (m instanceof Object) {\n" +
            "                msg = JSON.stringify(m);\n" +
            "            } else {\n" +
            "                msg = m.toString();\n" +
            "            }\n" +
            "        } else if (msgs.length > 1) {\n" +
            "            for (let i = 0; i < msgs.length; i++) {\n" +
            "                if (i > 0) {\n" +
            "                    msg = msg.concat(', ');\n" +
            "                }\n" +
            "                let m = msgs[i];\n" +
            "                if (typeof m === 'undefined') {\n" +
            "                    msg = msg.concat('undefined');\n" +
            "                } else if (m == null) {\n" +
            "                    msg = msg.concat('null');\n" +
            "                } else if (m instanceof Error) {\n" +
            "                    msg = msg.concat(m.toString() + '\\n' + m.stack);\n" +
            "                } else if (m instanceof Function) {\n" +
            "                    msg = msg.concat(m.toString());\n" +
            "                } else if (m instanceof Object) {\n" +
            "                    msg = msg.concat(JSON.stringify(m));\n" +
            "                } else {\n" +
            "                    msg = msg.concat(m.toString());\n" +
            "                }\n" +
            "            }\n" +
            "        }\n" +
            "        invoke(\"Hummer\", 0, funcName, msg);\n" +
            "    }\n" +
            "}\n" +
            "\n" +
            "const setTimeout = (func, timeout) => {\n" +
            "    let timer = new Timer();\n" +
            "    timer.setTimeout(func, timeout);\n" +
            "    return timer;\n" +
            "}\n" +
            "\n" +
            "const clearTimeout = (timer) => {\n" +
            "    if (timer instanceof Timer) {\n" +
            "        timer.clearTimeout();\n" +
            "    }\n" +
            "}\n" +
            "\n" +
            "const setInterval = (func, interval) => {\n" +
            "    let timer = new Timer();\n" +
            "    timer.setInterval(func, interval);\n" +
            "    return timer;\n" +
            "}\n" +
            "\n" +
            "const clearInterval = (timer) => {\n" +
            "    if (timer instanceof Timer) {\n" +
            "        timer.clearInterval();\n" +
            "    }\n" +
            "}\n" +
            "\n" +
            "const NotifyCenter = {\n" +
            "    addEventListener: (event, callback) => {\n" +
            "        invoke(\"NotifyCenter\", 0, \"addEventListener\", event, callback);\n" +
            "    },\n" +
            "    removeEventListener: (event, callback) => {\n" +
            "        invoke(\"NotifyCenter\", 0, \"removeEventListener\", event, callback);\n" +
            "    },\n" +
            "    triggerEvent: (event, value) => {\n" +
            "        invoke(\"NotifyCenter\", 0, \"triggerEvent\", event, JSON.stringify(value));\n" +
            "    }\n" +
            "}\n" +
            "\n" +
            "const Hummer = {\n" +
            "    setBasicWidth: (width) => {\n" +
            "        invoke(\"Hummer\", 0, \"setBasicWidth\", width);\n" +
            "    },\n" +
            "    render: (view) => {\n" +
            "        invoke(\"Hummer\", 0, \"render\", view.objID);\n" +
            "    },\n" +
            "    onRenderFinished: (isSucceed) => {\n" +
            "        invoke(\"Hummer\", 0, \"onRenderFinished\", isSucceed);\n" +
            "    },\n" +
            "    getRootView: () => {\n" +
            "        return invoke(\"Hummer\", 0, \"getRootView\");\n" +
            "    },\n" +
            "    loadScript: (script) => {\n" +
            "        return invoke(\"Hummer\", 0, \"loadScript\", script);\n" +
            "    },\n" +
            "    loadScriptWithUrl: (url, callback) => {\n" +
            "        invoke(\"Hummer\", 0, \"loadScriptWithUrl\", url, callback);\n" +
            "    },\n" +
            "    postException: (err) => {\n" +
            "        err = transSingleArg(err);\n" +
            "        invoke(\"Hummer\", 0, \"postException\", err);\n" +
            "    },\n" +
            "    notifyCenter: NotifyCenter,\n" +
            "}\n" +
            "\n" +
            "class Base {\n" +
            "    constructor(className, ...args) {\n" +
            "        this.className = className;\n" +
            "        this.objID = idGenerator();\n" +
            "        this.recycler = new Recycler(this.objID);\n" +
            "\n" +
            "        let params = transArgs(...args);\n" +
            "        invoke(this.className, this.objID, \"constructor\", this, ...params);\n" +
            "\n" +
            "        // 已弃用\n" +
            "        this.initialize(...args);\n" +
            "\n" +
            "        // 此方法只用于调试，为了统计组件树和函数调用树\n" +
            "        if (__IS_DEBUG__) {\n" +
            "            invoke(this.className, this.objID, \"constructor_end\", this);\n" +
            "        }\n" +
            "    }\n" +
            "\n" +
            "    // 已弃用\n" +
            "    initialize(...args) {}\n" +
            "\n" +
            "    set style(arg) {\n" +
            "        this._style = arg;\n" +
            "        arg = transSingleArg(arg);\n" +
            "        invoke(this.className, this.objID, \"setStyle\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get style() {\n" +
            "        return this._style;\n" +
            "    }\n" +
            "\n" +
            "    set enabled(arg) {\n" +
            "        this._enabled = arg;\n" +
            "        invoke(this.className, this.objID, \"setEnabled\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get enabled() {\n" +
            "        return this._enabled;\n" +
            "    }\n" +
            "\n" +
            "    set accessible(arg) {\n" +
            "        this._accessible = arg;\n" +
            "        invoke(this.className, this.objID, \"setAccessible\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get accessible() {\n" +
            "        return this._accessible;\n" +
            "    }\n" +
            "\n" +
            "    set accessibilityLabel(arg) {\n" +
            "        this._accessibilityLabel = arg;\n" +
            "        invoke(this.className, this.objID, \"setAccessibilityLabel\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get accessibilityLabel() {\n" +
            "        return this._accessibilityLabel;\n" +
            "    }\n" +
            "\n" +
            "    set accessibilityHint(arg) {\n" +
            "        this._accessibilityHint = arg;\n" +
            "        invoke(this.className, this.objID, \"setAccessibilityHint\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get accessibilityHint() {\n" +
            "        return this._accessibilityHint;\n" +
            "    }\n" +
            "\n" +
            "    set accessibilityRole(arg) {\n" +
            "        this._accessibilityRole = arg;\n" +
            "        invoke(this.className, this.objID, \"setAccessibilityRole\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get accessibilityRole() {\n" +
            "        return this._accessibilityRole;\n" +
            "    }\n" +
            "\n" +
            "    set accessibilityState(arg) {\n" +
            "        this._accessibilityState = arg;\n" +
            "        arg = transSingleArg(arg);\n" +
            "        invoke(this.className, this.objID, \"setAccessibilityState\", arg);\n" +
            "    }\n" +
            "\n" +
            "    get accessibilityState() {\n" +
            "        return this._accessibilityState;\n" +
            "    }\n" +
            "\n" +
            "    addEventListener(...args) {\n" +
            "        invoke(this.className, this.objID, \"addEventListener\", ...args);\n" +
            "    }\n" +
            "\n" +
            "    removeEventListener(...args) {\n" +
            "        invoke(this.className, this.objID, \"removeEventListener\", ...args);\n" +
            "    }\n" +
            "\n" +
            "    addAnimation(...args) {\n" +
            "        let stash = args;\n" +
            "        args = transArgs(...args);\n" +
            "        invoke(this.className, this.objID, \"addAnimation\", ...args);\n" +
            "    }\n" +
            "\n" +
            "    removeAnimationForKey(arg) {\n" +
            "        invoke(this.className, this.objID, \"removeAnimationForKey\", arg);\n" +
            "    }\n" +
            "\n" +
            "    removeAllAnimation() {\n" +
            "        invoke(this.className, this.objID, \"removeAllAnimation\");\n" +
            "    }\n" +
            "\n" +
            "    getRect(arg) {\n" +
            "        invoke(this.className, this.objID, \"getRect\", arg);\n" +
            "    }\n" +
            "\n" +
            "    resetStyle() {\n" +
            "        invoke(this.className, this.objID, \"resetStyle\");\n" +
            "    }\n" +
            "\n" +
            "    recycle() {\n" +
            "        invoke(this.className, this.objID, \"recycle\");\n" +
            "    }\n" +
            "\n" +
            "    dbg_highlight(arg) {\n" +
            "        let stash = arg;\n" +
            "        arg = transSingleArg(arg);\n" +
            "        invoke(this.className, this.objID, \"dbg_highlight\", arg);\n" +
            "    }\n" +
            "\n" +
            "    dbg_getDescription(...args) {\n" +
            "        let stash = args;\n" +
            "        args = transArgs(...args);\n" +
            "        invoke(this.className, this.objID, \"dbg_getDescription\", ...args);\n" +
            "    }\n" +
            "}\n" +
            "\n" +
            "__GLOBAL__.Hummer = Hummer;\n" +
            "__GLOBAL__.Base = Base;";

    /**
     * ES5版本的核心定义（仅用于Hermes调式模式）
     */
    public static String ES5_CORE = "";

    /**
     * ES5版本的SDK定义（仅用于Hermes调式模式）
     */
    public static String ES5_SDK = "";

    /**
     * ES5版本的组件定义（仅用于Hermes调式模式）
     */
    public static String ES5_COMP = "";

    /**
     * BABEL定义（仅用于Hermes调式模式）
     */
    public static String BABEL = "";
}
